torch/_inductor/freezing.py

```diff
@@ -0,0 +1,208 @@
+    gm.recompile()
+    return preserved_arg_indices
+
+
+@torch.utils._python_dispatch._disable_current_modes()
+def constant_fold(gm):
+    unknown_value = object()
+
+    node_replacements = {}
+
+    class ConstantFolder(torch.fx.Interpreter):
+        def run_node(self, node):
+            args, kwargs = self.fetch_args_kwargs_from_env(node)
+            if unknown_value in pytree.tree_flatten((args, kwargs))[0]:
+                return unknown_value
+
+            # All mutations should either be removed or on inputs which we did not make constant
+            if (
+                isinstance(node.target, torch._ops.OpOverload)
+                and torch.Tag.nondeterministic_seeded in node.target.tags
+            ):
+                return unknown_value
+
+            out = super().run_node(node)
+
+            # TODO - remove constant from node_replacement when it has no uses
+            if node.op != "get_attr" and isinstance(out, torch.Tensor):
+                node_replacements[node] = out
+
+            return out
+
+        def run(self):
+            env = {}
+            for n in self.module.graph.nodes:
+                if n.op == "placeholder":
+                    env[n] = unknown_value
+            return super().run(initial_env=env)
+
+    ConstantFolder(gm).run()
+
+    for node, constant in node_replacements.items():
+        replace_node_with_constant(gm, node, constant)
+
+    erased_params = []
+    for node in gm.graph.nodes:
+        if node.op == "get_attr" and len(node.users) == 0:
+            delattr(gm, node.target)
+            erased_params.append(node)
+
+    for node in erased_params:
+        gm.graph.erase_node(node)
+
+    gm.graph.eliminate_dead_code()
+    gm.graph.lint()
+    gm.recompile()
+
+
+def freeze(
+    dynamo_gm: torch.fx.GraphModule,
+    aot_autograd_gm: torch.fx.GraphModule,
+    fw_metadata,
+) -> Tuple[torch.fx.GraphModule, List[int]]:
+    """
+    Inlines parameters that are not mutated into constants and optimizes the graph through constant propagation
+    and other techniques. If enabled, the function also discards the original parameters of the module for memory efficiency.
+
+    Assumes that this function is run in dynamo tracing post aot_autograd.
+
+    Args:
+        dynamo_gm (torch.fx.GraphModule): The Dynamo constructed GraphModule.
+        aot_autograd_gm (torch.fx.GraphModule): The aot_autograd constructed GraphModule to be frozen.
+        example_inputs_ (List[torch.Tensor]): A list of example input tensors to be used in the freezing process.
+        fw_metadata: Metadata for the forward method of the graph module.
+
+    Returns:
+        Tuple[torch.fx.GraphModule, List[int]]: A tuple containing the frozen GraphModule and a list of indices
+        of the inputs that were preserved (not turned into constants).
+    """
+    params_flat = torch._guards.TracingContext.get().params_flat
+    preserved_arg_indices = replace_params_with_constants(
```

Comment:
another curiosity question - do you imagine that shape padding of parameters would also (maybe) happen here?

Reply:
It happens naturally as part of the constant folding